teuthology/suite/merge.py

```diff
@@ -0,0 +1,273 @@
+import copy
+import logging
+import lupa
+import yaml
+
+from teuthology.suite.build_matrix import combine_path
+from teuthology.suite.util import strip_fragment_path
+from teuthology.misc import deep_merge
+
+log = logging.getLogger(__name__)
+
+TEUTHOLOGY_TEMPLATE = {
```

Comment:
If anything ever modifies this value, we'll start to see some very mysterious issues pop up. Nothing in the PR will modify it thanks to your modifications to `deep_merge()`, but we can take an extra step to avoid future issues by making this `dict` a [`MappingProxyType`](https://docs.python.org/3/library/types.html#types.MappingProxyType). That's a trivial change, and the only other one required would be to add `or isinstance(b, MappingProxyType)` to the `assert` in `deep_merge()`.

Reply:
Makes sense!